bundles/org.openhab.binding.e3dc/src/main/java/org/openhab/binding/e3dc/internal/handler/E3DCDeviceThingHandler.java

```diff
@@ -0,0 +1,178 @@
+ *
+ * @author Bernd Weymann - Initial contribution
+ */
+@NonNullByDefault
+public class E3DCDeviceThingHandler extends BaseBridgeHandler
+        implements DataListener, ModbusWriteCallback, ModbusFailureCallback<ModbusWriteRequestBlueprint> {
+    private final Logger logger = LoggerFactory.getLogger(E3DCDeviceThingHandler.class);
+    private ModbusManager modbusManagerRef;
+    private final ModbusCallback modbusInfoCallback = new ModbusCallback(DataType.INFO);
+    private final ModbusCallback modbusDataCallback = new ModbusCallback(DataType.DATA);
+    private ThingStatus myStatus = ThingStatus.UNKNOWN;
+    private @Nullable ModbusCommunicationInterface modbusCom;
+    private @Nullable PollTask infoPoller;
+    private @Nullable PollTask dataPoller;
+    private @Nullable E3DCDeviceConfiguration config;
+
+    public E3DCDeviceThingHandler(Bridge bridge, ModbusManager ref) {
+        super(bridge);
+        modbusManagerRef = ref;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+    }
+
+    @Override
+    public void initialize() {
+        setStatus(ThingStatus.UNKNOWN);
+        // Example for background initialization:
+        scheduler.execute(() -> {
+            config = getConfigAs(E3DCDeviceConfiguration.class);
+            E3DCDeviceConfiguration localConfig = config;
+            if (localConfig != null && checkConfig(localConfig)) {
+                ModbusTCPSlaveEndpoint slaveEndpoint = new ModbusTCPSlaveEndpoint(localConfig.host, localConfig.port);
+
+                ModbusCommunicationInterface localModbusCom = modbusManagerRef.newModbusCommunicationInterface(
+                        slaveEndpoint, modbusManagerRef.getEndpointPoolConfiguration(slaveEndpoint));
+                // register low speed info poller
+                ModbusReadRequestBlueprint infoRequest = new ModbusReadRequestBlueprint(localConfig.deviceid,
+                        ModbusReadFunctionCode.READ_MULTIPLE_REGISTERS, INFO_REG_START, INFO_REG_SIZE, 3);
+                infoPoller = localModbusCom.registerRegularPoll(infoRequest, INFO_POLL_REFRESH_TIME_MS, 0,
+                        modbusInfoCallback, modbusInfoCallback);
+
+                ModbusReadRequestBlueprint dataRequest = new ModbusReadRequestBlueprint(localConfig.deviceid,
+                        ModbusReadFunctionCode.READ_MULTIPLE_REGISTERS, POWER_REG_START,
+                        REGISTER_LENGTH - INFO_REG_SIZE, 3);
+                dataPoller = localModbusCom.registerRegularPoll(dataRequest, localConfig.refresh, 0, modbusDataCallback,
+                        modbusDataCallback);
+                modbusCom = localModbusCom;
+                // listen for data to get ONLINE
+                modbusDataCallback.addDataListener(this);
+            } else {
+                setStatus(ThingStatus.OFFLINE);
+            }
+        });
+    }
+
+    @Override
+    public void dispose() {
```

Comment:
In case you decide not to follow my recommendation of reusing `tcp`/`serial` things from the modbus binding, you need to ensure yourself that `ModbusCommunicationInterface` is closed. This is normally done by the endpoint thing: https://github.com/openhab/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/handler/AbstractModbusEndpointThingHandler.java#L98

Reply:
Obsolete due to change towards Modbus binding